### PR TITLE
Updated the example for the private eks cluster with additional details and individual stacks for vpc, eks cluster and add-ons

### PR DESCRIPTION
**Pull request by satveerkhurpa**
### What does this PR do?
This PR updates the example to provision a private eks cluster. The updates includes individual terraform stacks to set up the VPC, the eks cluster and add-ons. also includes a high level design that shows the users on how they can use a jenkins host or a bastion host from the default vpc, a peering connection between the default vpc and the eks vpc to complete the set up of the private eks cluster.

### Motivation

Users will have an example to provision a private eks cluster the right way.

### More

- [ x] Yes, I have tested the PR using my local account setup  (Provide any test evidence report under Additional Notes)
- [ x] Yes, I have added a new example under [examples](https://github.com/aws-samples/aws-eks-accelerator-for-terraform/tree/main/examples) to support my PR
- [x ] Yes, I have created another PR for add-ons under [add-ons](https://github.com/aws-samples/ssp-eks-add-ons) repo (if applicable)
- [x ] Yes, I have updated the [docs](https://github.com/aws-samples/aws-eks-accelerator-for-terraform/tree/main/docs) for this feature
- [ ] Yes, I ran `pre-commit run -a` with this PR

**Note**: Not all the PRs required examples and docs except a new pattern or add-on added.

### For Moderators
- [ ] E2E Test successfully complete before merge?

### CODE REVIEW

File: examples/fully-private-eks-cluster/README.md [hunk 1]
```diff
@@ -1,139 +1,45 @@
...
-```
-
-Enter `yes` to apply
-
-### Configure `kubectl` and test cluster
-EKS Cluster details can be extracted from terraform output or from AWS Console to get the name of cluster.
-This following command used to update the `kubeconfig` in your local machine where you run kubectl commands to interact with your EKS Cluster.
-
-#### Step5: Run `update-kubeconfig` command
+Here is the high level design of the solution. The solution has been split into 3 different TerraForm stacks for simplicity.
```
Comment (kcoleman731) [inline examples/fully-private-eks-cluster/README.md:10]: Terraform

File: examples/fully-private-eks-cluster/add-ons/main.tf [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+  enable_ingress_nginx                = false
+  enable_karpenter                    = true
+  enable_keda                         = true
+  enable_metrics_server               = true
+  enable_prometheus                   = false
+  enable_traefik                      = false
+  enable_vpa                          = true
+  enable_yunikorn                     = false
+  enable_argo_rollouts                = false
+
```
Comment (kcoleman731) [inline examples/fully-private-eks-cluster/add-ons/main.tf:115]: extra line

File: examples/fully-private-eks-cluster/add-ons/main.tf [hunk 3]
```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (kcoleman731) [inline examples/fully-private-eks-cluster/add-ons/main.tf:2]: Copyright  headers not needed

File: examples/fully-private-eks-cluster/eks/main.tf [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+provider "helm" {
+  kubernetes {
+    host                   = data.aws_eks_cluster.cluster.endpoint
+    token                  = data.aws_eks_cluster_auth.cluster.token
+    cluster_ca_certificate = base64decode(data.aws_eks_cluster.cluster.certificate_authority.0.data)
+  }
+}
+
+data "aws_region" "current" {}
```
Comment (kcoleman731) [inline examples/fully-private-eks-cluster/eks/main.tf:63]: put in a data.tf file for consistency

File: examples/fully-private-eks-cluster/eks/main.tf [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+  kubernetes_version = var.cluster_version
+  terraform_version  = "Terraform v1.0.1"
+
+  vpc_id             = data.terraform_remote_state.vpc_s3_backend.outputs.vpc_id
+  private_subnet_ids = data.terraform_remote_state.vpc_s3_backend.outputs.private_subnets
+  public_subnet_ids  = data.terraform_remote_state.vpc_s3_backend.outputs.public_subnets
+
+  cluster_name = join("-", [local.tenant, local.environment, local.zone, "eks"])
+
```
Comment (kcoleman731) [inline examples/fully-private-eks-cluster/eks/main.tf:98]: extran line

File: examples/fully-private-eks-cluster/eks/main.tf [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+  # EKS MANAGED NODE GROUPS
+  managed_node_groups = {
+    mg_4 = {
+      node_group_name = "managed-ondemand"
+      instance_types  = ["m5.xlarge"]
+      subnet_ids      = local.private_subnet_ids
+    }
+  }
+}
```
Comment (kcoleman731) [inline examples/fully-private-eks-cluster/eks/main.tf:131]: add a new line

File: examples/fully-private-eks-cluster/eks/variables.tf [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+  default     = []
+}
+
+variable "cluster_security_group_additional_rules" {
+  description = "List of additional security group rules to add to the cluster security group created. Set `source_node_security_group = true` inside rules to set the `node_security_group` as source"
+  type        = any
+  default     = {}
+}
+
+
```
Comment (kcoleman731) [inline examples/fully-private-eks-cluster/eks/variables.tf:70]: extra line

File: examples/fully-private-eks-cluster/vpc/main.tf [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+  vpc_id = module.aws_vpc.vpc_id
+}
+
+module "vpc_endpoints" {
+  source  = "terraform-aws-modules/vpc/aws//modules/vpc-endpoints"
+  version = "v3.2.0"
+  create  = true
+  vpc_id  = module.aws_vpc.vpc_id
+  security_group_ids = [
+  data.aws_security_group.default.id]
```
Comment (kcoleman731) [inline examples/fully-private-eks-cluster/vpc/main.tf:123]: format

File: examples/fully-private-eks-cluster/vpc/outputs.tf [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+output "private_subnets" {
+  description = "List of IDs of private subnets"
+  value       = module.aws_vpc.private_subnets
+}
+
+output "public_subnets" {
+  description = "List of IDs of public subnets"
+  value       = module.aws_vpc.public_subnets
+}
```
Comment (kcoleman731) [inline examples/fully-private-eks-cluster/vpc/outputs.tf:32]: new line

**Review by kcoleman731 — COMMENTED** [review]
Looking great some nit comments for you. Please connect with @Zvikan on adding this into e2e tests.

**Comment by satveerkhurpa** [comment]
> Looking great some nit comments for you. Please connect with @Zvikan on adding this into e2e tests.

@kcoleman731 - Pushed updates based on review comments.

File: examples/fully-private-eks-cluster/vpc/backend.conf [hunk 10]
```diff
@@ -0,0 +1,3 @@
+bucket = "terraform-ssp-github-actions-state" #Create a new S3 bucket for terraform state management and update this entry with the bucket details
+region = "us-west-2"
+key    = "private/vpc/terraform-main.tfstate"
```
Comment (vara-bonthu) [inline examples/fully-private-eks-cluster/vpc/backend.conf:3]: We don't want users to these values. Ask the users to enter the bucket and ket details and add comments

File: examples/fully-private-eks-cluster/vpc/main.tf [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+ * permit persons to whom the Software is furnished to do so.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
```
Comment (vara-bonthu) [inline examples/fully-private-eks-cluster/vpc/main.tf:18]: remove this

File: examples/fully-private-eks-cluster/vpc/main.tf [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+module "aws_vpc" {
+  source  = "terraform-aws-modules/vpc/aws"
+  version = "v3.2.0"
+
+  name = local.vpc_name
+  cidr = local.vpc_cidr
+  azs  = data.aws_availability_zones.available.names
+
+  public_subnets  = [for k, v in slice(data.aws_availability_zones.available.names, 0, 3) : cidrsubnet(local.vpc_cidr, 8, k)]
```
Comment (vara-bonthu) [inline examples/fully-private-eks-cluster/vpc/main.tf:50]: remove this and just show private subnets only

File: examples/fully-private-eks-cluster/vpc/outputs.tf [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+ * permit persons to whom the Software is furnished to do so.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
```
Comment (vara-bonthu) [inline examples/fully-private-eks-cluster/vpc/outputs.tf:18]: remove this

File: examples/fully-private-eks-cluster/vpc/outputs.tf (hunk 9 above)
Comment (vara-bonthu) [inline examples/fully-private-eks-cluster/vpc/outputs.tf:32]: no public subnets for this pattern

File: examples/fully-private-eks-cluster/vpc/variables.tf [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+ * merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
+ * permit persons to whom the Software is furnished to do so.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
```
Comment (vara-bonthu) [inline examples/fully-private-eks-cluster/vpc/variables.tf:17]: remove this

**Review by vara-bonthu — COMMENTED** [review]
Thanks @satveerkhurpa ! few comments added.

I think we need to show this example without using any public subnets. It would nice to show the add-ons using the ECR repo for docker images and helm charts.

File: examples/fully-private-eks-cluster/eks/base.tfvars [hunk 15]
```diff
@@ -0,0 +1,14 @@
+region             = "us-west-2"
+cluster_version    = "1.22"
+vpc_id             = "vpc-045cde3975b801469"
+private_subnet_ids = ["subnet-03a21a869766169d5", "subnet-0899effaed018015c", "subnet-085209c3a039361b1"]
```
Comment (vara-bonthu) [inline examples/fully-private-eks-cluster/eks/base.tfvars:4]: remove the actual values and replace this with `"<ENTER_VPC_ID>"`

File: examples/fully-private-eks-cluster/vpc/base.tfvars [hunk 16]
```diff
@@ -0,0 +1,3 @@
+region                = "us-west-2"
+vpc_cidr              = "10.0.0.0/16"
+default_vpc_ipv4_cidr = "172.31.0.0/16"
```
Comment (vara-bonthu) [inline examples/fully-private-eks-cluster/vpc/base.tfvars:3]: Do you need `default_vpc_ipv4_cidr` ?

File: examples/fully-private-eks-cluster/vpc/variables.tf [hunk 17]
```diff
@@ -0,0 +1,17 @@
+variable "region" {
+  type        = string
+  description = "AWS region"
+  default     = "us-east-1"
```
Comment (vara-bonthu) [inline examples/fully-private-eks-cluster/vpc/variables.tf:4]: I can see us-west-2 also used in this example. Can we make the region consistent across the example?

File: examples/fully-private-eks-cluster/eks/main.tf [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+  vpc_id             = local.vpc_id
+  private_subnet_ids = local.private_subnet_ids
+
+  # Cluster Security Group
+  cluster_security_group_additional_rules = var.cluster_security_group_additional_rules
+
+  # EKS CONTROL PLANE VARIABLES
+  cluster_version = var.cluster_version
+
+  cluster_endpoint_public_access  = false
```
Comment (vara-bonthu) [inline examples/fully-private-eks-cluster/eks/main.tf:43]: Change this to `true` and add notes to changed to `false` once the cluster is deployed. Also, advise users to use Bastion host to communicate with cluster once its changed to false

File: examples/fully-private-eks-cluster/eks/base.tfvars [hunk 19]
```diff
@@ -0,0 +1,14 @@
...
+cluster_security_group_additional_rules = {
+  ingress_from_jenkins_host = {
+    description = "Ingress from Jenkins/Bastion Host"
+    protocol    = "-1"
+    from_port   = 0
+    to_port     = 0
+    type        = "ingress"
+    cidr_blocks = ["172.31.0.0/16"]
```
Comment (vara-bonthu) [inline examples/fully-private-eks-cluster/eks/base.tfvars:12]: Are we creating a Bastion host in this example? If not provide instructions to create a Bastion host in another VPC and steps to VPC peering

**Review by vara-bonthu — COMMENTED** [review]
We also need to show at-least two Add-ons deployed using ECR. Here we update the Helm chart repo URL and Image config pointing to ECR

File: examples/fully-private-eks-cluster/add-ons/main.tf [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+    host                   = data.aws_eks_cluster.cluster.endpoint
+    token                  = data.aws_eks_cluster_auth.cluster.token
+    cluster_ca_certificate = base64decode(data.aws_eks_cluster.cluster.certificate_authority.0.data)
+  }
+}
+
+locals {
+
+  cluster_version = var.cluster_version
+
```
Comment (askulkarni2) [inline examples/fully-private-eks-cluster/add-ons/main.tf:26]: nit: extra newlines

File: examples/fully-private-eks-cluster/add-ons/versions.tf [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+  }
+
+
+  #State managed via Terraform Cloud
+  cloud {
+    organization = "skdemo"
+    workspaces {
+      name = "private-addons"
+    }
+  }
```
Comment (askulkarni2) [inline examples/fully-private-eks-cluster/add-ons/versions.tf:26]: remove this

File: examples/fully-private-eks-cluster/eks/versions.tf [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+    }
+  }
+
+  cloud {
+    organization = "skdemo"
+
+    workspaces {
+      name = "private-eks"
+    }
+  }
```
Comment (askulkarni2) [inline examples/fully-private-eks-cluster/eks/versions.tf:25]: please remove

File: examples/fully-private-eks-cluster/vpc/versions.tf [hunk 23]
```diff
@@ -16,10 +16,18 @@ terraform {
...
   }
 
+  cloud {
+    organization = "skdemo"
+
+    workspaces {
+      name = "private-vpc"
+    }
+  }
+
```
Comment (askulkarni2) [inline examples/fully-private-eks-cluster/vpc/versions.tf:26]: please remove

File: examples/fully-private-eks-cluster/vpc/vpc-peer.tf [hunk 24]
```diff
@@ -0,0 +1,54 @@
...
+  route_table_id            = element(module.aws_vpc.private_route_table_ids, count.index)
+  destination_cidr_block    = local.cloud9_vpc_cidr
+  vpc_peering_connection_id = aws_vpc_peering_connection.cloud9-eks-peer.id
+}
+
+
+# resource "aws_iam_instance_profile" "eks-cloud9_instance_profile" {
+#   name = "eks-cloud9_instance_profile"
+#   role = local.cloud9_iam_role_arn
+# }
```
Comment (askulkarni2) [inline examples/fully-private-eks-cluster/vpc/vpc-peer.tf:36]: remove if this is not used

File: examples/multi-tenancy-with-teams/main.tf [hunk 25]
```diff
@@ -46,7 +46,8 @@ data "aws_caller_identity" "current" {}
 data "aws_availability_zones" "available" {}
 
 locals {
-  name   = basename(path.cwd)
+  #name   = basename(path.cwd)
+  name   = "multitenancy-eks"
```
Comment (askulkarni2) [inline examples/multi-tenancy-with-teams/main.tf:50]: why this change?

File: examples/multi-tenancy-with-teams/versions.tf [hunk 26]
```diff
@@ -20,6 +20,14 @@ terraform {
...
   }
 
+  cloud {
+    organization = "skdemo"
+
+    workspaces {
+      name = "multi-tenancy-eks"
+    }
+  }
+
```
Comment (askulkarni2) [inline examples/multi-tenancy-with-teams/versions.tf:30]: please remove

File: modules/irsa/main.tf [hunk 27]
```diff
@@ -38,8 +38,9 @@ resource "aws_iam_role" "irsa" {
         },
         "Action" : "sts:AssumeRoleWithWebIdentity",
         "Condition" : {
-          "StringLike" : {
-            "${var.addon_context.eks_oidc_issuer_url}:sub" : "system:serviceaccount:${var.kubernetes_namespace}:${var.kubernetes_service_account}"
+          "StringEquals" : {
+            "${var.addon_context.eks_oidc_issuer_url}:sub" : "system:serviceaccount:${var.kubernetes_namespace}:${var.kubernetes_service_account}",
+            "${var.addon_context.eks_oidc_issuer_url}:aud" : "sts.amazonaws.com"
```
Comment (askulkarni2) [inline modules/irsa/main.tf:43]: @satveerkhurpa how did you arrive at this change? I want to understand possible ramifications of this.

cc: @vara-bonthu, @Zvikan, @bryantbiggs

**Review by askulkarni2 — COMMENTED** [review]
We will likely have to skip this example from our parallel-plans CI check as this example does not conform to the structure that's expected for that. But as we plan to iterate on this example going forward we can look at those changes down the road. I also want to validate the change to the IRSA module. But other than that (and CI failures on `pre-commit`) this is looking good to me.

File: modules/irsa/main.tf (hunk 27 above)
Comment (bryantbiggs) [reply modules/irsa/main.tf:43]: this is similar to what we have upstream https://github.com/terraform-aws-modules/terraform-aws-iam/blob/master/modules/iam-role-for-service-accounts-eks/main.tf#L16-L26

File: examples/fully-private-eks-cluster/add-ons/README.md [hunk 28]
```diff
@@ -0,0 +1,116 @@
...
+without restriction, including without limitation the rights to use, copy, modify,
+merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
+permit persons to whom the Software is furnished to do so.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
```
[the remark this answers is not in the document]
Comment (satveerkhurpa) [reply examples/fully-private-eks-cluster/add-ons/README.md]: The individual module README.md were removed.

File: examples/fully-private-eks-cluster/add-ons/backend.conf [hunk 29]
```diff
@@ -0,0 +1,3 @@
+bucket = "terraform-ssp-github-actions-state"
+region = "us-west-2"
+key    = "private/addons/terraform-main.tfstate"
```
[the remark this answers is not in the document]
Comment (satveerkhurpa) [reply examples/fully-private-eks-cluster/add-ons/backend.conf]: This has been cleaned up quite some time back and may have been an old remnant.

File: examples/fully-private-eks-cluster/add-ons/base.tfvars [hunk 30]
```diff
@@ -0,0 +1,2 @@
+region = "us-west-2"
```
[the remark this answers is not in the document]
Comment (satveerkhurpa) [reply examples/fully-private-eks-cluster/add-ons/base.tfvars]: all tfvar file were removed.

File: modules/irsa/main.tf (hunk 27 above)
Comment (satveerkhurpa) [reply modules/irsa/main.tf:36]: Yes, I verified the upstream codebase and updated accordingly.

File: examples/fully-private-eks-cluster/vpc/vpc-peer.tf [hunk 31]
```diff
@@ -0,0 +1,47 @@
...
+  owner_arn     = local.cloud9_owner_arn
+}
+
+# data "aws_instance" "cloud9_instance" {
+#   filter {
+#     name = "tag:aws:cloud9:environment"
+#     values = [
+#     aws_cloud9_environment_ec2.eks-cloud9.id]
+#   }
+# }
```
Comment (askulkarni2) [inline examples/fully-private-eks-cluster/vpc/vpc-peer.tf:47]: Remove commented block if not needed.

**Review by askulkarni2 — COMMENTED** [review]
Please address the minor comments otherwise its ready to go.

File: examples/multi-tenancy-with-teams/versions.tf (hunk 26 above)
Comment (satveerkhurpa) [reply examples/multi-tenancy-with-teams/versions.tf:30]: Done

File: examples/multi-tenancy-with-teams/main.tf (hunk 25 above)
Comment (satveerkhurpa) [reply examples/multi-tenancy-with-teams/main.tf:50]: Rolled back. Was testing something.

**Review by askulkarni2 — APPROVED** [review]
LGTM! Thanks for the PR @satveerkhurpa.
